src/opt/Optimizer/TypedDesugarer.hs

```diff
@@ -0,0 +1,257 @@
+          m{mbody = desugarExpr (mbody m)}
+
+      desugarExpr ast =
+          foldl (\ast opt -> opt ast) ast desugarPasses
+
+-- | The functions in this list will be performed in order during desugaring
+desugarPasses :: [Expr -> Expr]
+desugarPasses = [desugarAndBoxForInSeq, desugarAndBoxForNotInSeq]
+
+desugarAndBoxForInSeq = extend desugarAndBoxForInSeq'
+  where desugarAndBoxForInSeq' e@Seq{} = boxed e
+        desugarAndBoxForInSeq' e = e
+        boxed e = boxForInSeq e
+
+desugarAndBoxForNotInSeq = extend desugarAndBoxForNotInSeq'
+  where desugarAndBoxForNotInSeq' for@For{} = forBoxed for []
+        desugarAndBoxForNotInSeq' e = e
+
+
+
+-- Desugars a for-loop into nested calls to map and flatMap and foreach:
+--
+-- for x <- listA, y <- listB, z <- ListC do
+--      fun
+-- end
+--
+-- into listA.flatMap(listB.flatMap(listC.map(fun)))
+forDesugared :: Expr -> Expr
+forDesugared e@For{emeta, sources, body} =
+  let n = length sources
+      collectionType = getType $ collection $ head sources
+      callNameList = if (not (AST.AST.isCaptured e)) || (unitType == getType body) || (isRangeObjectType collectionType)
+                     then replicate n (Name "foreach")
+                     else replicate (n-1) (Name "flatMap") ++ [Name "map"]
+      revSources = reverse sources
+      elemType = getType body
+      desugaredFor = nestCalls emeta callNameList sources body elemType
+  in  desugaredFor
+forDesugared m = m
+
+nestCalls :: Meta.Meta Expr -> [Name] -> [ForSource] -> Expr -> Type -> Expr
+nestCalls meta (name:_) (fs:[]) body elemType = intoCall meta name fs body elemType
+nestCalls meta (name:restOfNames) (fs:restFS) body elemType =
+  let nestedCall = intoCall meta name fs body elemType
+  in nestCalls meta restOfNames restFS nestedCall elemType
+
+intoCall :: Meta.Meta Expr -> Name -> ForSource -> Expr -> Type -> Expr
+intoCall met callName ForSource{fsName, fsTy, collection} bodyOrMethodCall elemType =
+  if isRefType (getType collection)
+  then let param = [intoParam met Val fsName fsTy]
+           arguments = [intoClosure met param Nothing bodyOrMethodCall]
+           elemT = if callName == Name "foreach" -- this feels iffy
+                   then []
+                   else [elemType]
+       in  intoMethodCall met elemT collection callName arguments
+   else let param = [intoParam met Val fsName fsTy]
+            arguments = [intoClosure met param Nothing bodyOrMethodCall] ++ [collection]
+            elemT =  if callName == Name "foreach" -- this feels iffy
+                     then [fromMaybe intType fsTy]
+                     else [(fromMaybe intType fsTy), elemType]
+            name = intoQName callName
+       in   intoFunctionCall met elemT name arguments
+
+boxForInSeq :: Expr -> Expr
+boxForInSeq e@Seq{emeta, eseq} =
+  let newEseq = boxFor eseq []
+  in  e{eseq = newEseq}
+  where
+    boxFor :: [Expr] -> [Expr] -> [Expr]
+    boxFor [] newEseq =  newEseq
+    boxFor (ex:expr) newEseq
```

Comment:
I believe this code is the same as:
```
boxFor [] newEseq                   =  newEseq
boxFor ((For{}):expr) newEseq = newEseq ++ [forBoxed ex (boxFor expr [])]
boxFor (ex:expr) newEseq        = boxFor expr (newEseq ++ [ex])
```
Maybe this is more readable than calling that function.